### PR TITLE
core: bug in alloc_samples ?

### PR DESCRIPTION
**Pull request by janiversen**
When num > dc->alloc_samples we whould allocate space for num

Signed-off-by: Jan Iversen <jani@apache.org>

### Describe the pull request:

- [ x] Bug fix
- [ ] Functional change
- [ ] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

alloc_samples does not use the parameter num, and dc->alloc_samples start with 0,
so change to realloc using num (if > dc->alloc_samples). I saw this problem while debugging #1408

Please this code is completely new land, so please check if my assumptions are correct.

The second commit for qmlmanager.cpp corrects the problem in #1408, but I am not sure it
is the correct way.

solves #1408

### CODE REVIEW

**Comment by janiversen** [comment]
not enough, but still an error in my opinion

**Review by bstoeger — COMMENTED** [review]
lgtm - only a minor comment.

File: mobile-widgets/qmlmanager.cpp [hunk 1]
```diff
@@ -927,6 +927,7 @@ bool QMLManager::checkDuration(DiveObjectHelper *myDive, struct dive *d, QString
 			free(d->dc.sample);
 			d->dc.sample = 0;
 			d->dc.samples = 0;
+			d->dc.alloc_samples = 0;
```
Comment (bstoeger) [inline mobile-widgets/qmlmanager.cpp:930]: It is kind of sad that the UI-code directly manipulates internal data structures.

Therefore, in the short-run, I would suggest creating a `free_samples()` function, which does the above and is called here and in `planner.c`. But I think this is beyond the scope of this PR.

In the long run, I think we should reimplement `std::vector<>` in C [my personal choice would be simply using `std::vector<>` in the backend (while still providing a C-only interface). But no worries - I'm not going to try to convince anybody.]

Comment (janiversen) [reply mobile-widgets/qmlmanager.cpp:930]: I agree that making  a free_samples function is the minimum we should do, but I wanted to keep this PR as small as possible (as you also write out of scope). If/when this PR is merged I am happy to make the function.

I agree with you on the vector considerations.

File: core/dive.c [hunk 2]
```diff
@@ -747,7 +747,7 @@ void copy_samples(struct divecomputer *s, struct divecomputer *d)
 void alloc_samples(struct divecomputer *dc, int num)
 {
 	if (num > dc->alloc_samples) {
-		dc->alloc_samples = (dc->alloc_samples * 3) / 2 + 10;
+		dc->alloc_samples = (num * 3) / 2 + 10;
```
Comment (dirkhh) [inline core/dive.c:750]: So this allocates 50% + 10 additional samples. Which of course works fine if your pattern everywhere in the code is to either simply allocate "one more" sample (so this gets called with 'nr+1'), or up to six samples in one block.
That is of course bleeping fragile and needs to be fixed - but given our call pattern, I wonder how you got into trouble here?

File: mobile-widgets/qmlmanager.cpp (hunk 1 above)
Comment (dirkhh) [reply mobile-widgets/qmlmanager.cpp:930]: C++ code compiles slower and runs usually an order of magnitude slower. Which is why we generally don't jump on that bandwagon for things that can just as easily done in C.

**Review by dirkhh — COMMENTED** [review]
As I said in my comment on the other PR: please use
```
Fixes #1410
```
in the commit message which fixes that issue

File: core/dive.c (hunk 2 above)
Comment (janiversen) [reply core/dive.c:750]: just for the record, this did not bring trouble, but I saw it while debugging, and thought it looked wrong. I did not adjust the algorithm as such, the structure is not very big so overallocating is not a big issue.

**Pull request merged**